src/ULog.ts

```diff
@@ -220,37 +259,73 @@ export class ULog {
     return this.#dataTimeRange;
   }
 
-  async #createIndex(): Promise<void> {
+  async #createIndex(reader: ChunkedReader): Promise<void> {
+    if (!this.#header) {
+      throw new Error(`Cannot read before open`);
+    }
+
     const timeIndex: IndexEntry[] = [];
     const dataCounts = new Map<number, number>();
     let minTimestamp: bigint | undefined;
     let maxTimestamp = 0n;
     let logMessageCount = 0;
 
-    let i = 0;
-    let message: DataSectionMessage | undefined;
-    while ((message = await this.#readParsedMessage())) {
-      if (message.type === MessageType.Data) {
-        if (minTimestamp == undefined || message.value.timestamp < minTimestamp) {
-          minTimestamp = message.value.timestamp;
+    for (;;) {
+      const offset = reader.position();
+
+      const rawMessage = await readRawMessage(reader, this.#dataEnd);
+      if (!rawMessage) {
+        break;
+      }
+
+      if (rawMessage.type === MessageType.AddLogged) {
+        this.#handleSubscription(rawMessage);
+      }
+
+      if (rawMessage.type === MessageType.Log || rawMessage.type === MessageType.LogTagged) {
+        if (minTimestamp == undefined || rawMessage.timestamp < minTimestamp) {
+          minTimestamp = rawMessage.timestamp;
         }
-        if (message.value.timestamp > maxTimestamp) {
-          maxTimestamp = message.value.timestamp;
+        if (rawMessage.timestamp > maxTimestamp) {
+          maxTimestamp = rawMessage.timestamp;
         }
-        timeIndex.push([message.value.timestamp, i++, message]);
-        dataCounts.set(message.msgId, (dataCounts.get(message.msgId) ?? 0) + 1);
-      } else if (message.type === MessageType.Log || message.type === MessageType.LogTagged) {
-        if (minTimestamp == undefined || message.timestamp < minTimestamp) {
-          minTimestamp = message.timestamp;
+
+        timeIndex.push([rawMessage.timestamp, offset, undefined]);
+        logMessageCount++;
+        continue;
+      }
+
+      if (rawMessage.type === MessageType.Data) {
+        const dataMsg = rawMessage;
+        const definition = this.#subscriptions.get(dataMsg.msgId);
+        if (!definition) {
+          const msgPos = reader.position() - rawMessage.size - 3;
+          throw new Error(
+            `Unknown msg_id ${dataMsg.msgId} for ${rawMessage.size} byte 'D' message at offset ${msgPos}`,
```

Comment:
does `'D' message` mean something in ulog? This error message seems confusing - is it missing a word `in`?

Reply:
No idea.. it was present in the previous code https://github.com/foxglove/ulog/blob/main/src/ULog.ts#L285 and I didn't think much of it